galleon-plugins/src/main/java/org/wildfly/galleon/plugin/ShadedModel.java

```diff
@@ -0,0 +1,260 @@
+                rootElement.getNamespaceURI());
+        Elements elements = manifestEntries.getChildElements();
+        for (int i = 0; i < elements.size(); i++) {
+            Element e = elements.get(i);
+            entries.put(e.getLocalName(), e.getValue());
+        }
+        return entries;
+    }
+
+    public String getMainClass() {
+        String ret = null;
+        Element mainClass = rootElement.getFirstChildElement("main-class",
+                rootElement.getNamespaceURI());
+        if (mainClass != null) {
+            ret = mainClass.getValue();
+        }
+        return ret;
+    }
+
+    public String getName() {
+        return rootElement.getFirstChildElement("name",
+                rootElement.getNamespaceURI()).getValue();
+    }
+
+    public void buildJar(Path shadedJar) throws IOException, ProvisioningException {
+        if (log.isVerboseEnabled()) {
+            log.verbose("Assembling shaded jar " + shadedJar);
+        }
+        Path tmpTarget = tmpPath.resolve("assemble_target").resolve(shadedJar.getFileName());
+        Files.createDirectories(tmpTarget);
+        for (MavenArtifact dependency : getArtifacts()) {
+            Utils.navigateArtifact(dependency.getPath(), tmpTarget, this);
+        }
+        update(tmpTarget);
+        ZipUtils.zip(tmpTarget, shadedJar);
+        IoUtils.recursiveDelete(tmpTarget);
+    }
+
+    public void update(Path target) throws IOException {
+        Path targetMetaInf = target.resolve("META-INF");
+        Path targetManifestPath = targetMetaInf.resolve("MANIFEST.MF");
+        Manifest manifest;
+        if (!Files.exists(targetManifestPath)) {
+            manifest = new Manifest();
+        } else {
+            try (FileInputStream stream = new FileInputStream(targetManifestPath.toFile())) {
+                manifest = new Manifest(stream);
+            }
+        }
+        Attributes attributes = manifest.getMainAttributes();
+        String mainClass = getMainClass();
+        if (mainClass != null) {
+            attributes.put(Attributes.Name.MAIN_CLASS, mainClass);
+        }
+        Map<String, String> manifestEntries = getManifestEntries();
+        if (manifestEntries != null) {
+            for (Map.Entry<String, String> entry : manifestEntries.entrySet()) {
+                if (entry.getValue() == null) {
+                    attributes.remove(new Attributes.Name(entry.getKey()));
+                } else {
+                    attributes.put(new Attributes.Name(entry.getKey()), entry.getValue());
+                }
+            }
+        }
+        attributes.put(Attributes.Name.IMPLEMENTATION_TITLE, "Galleon shading of " + getName());
+        attributes.put(Attributes.Name.SPECIFICATION_TITLE, "Galleon shading of " + getName());
+        attributes.put(Attributes.Name.IMPLEMENTATION_VERSION, "Unknown");
+        Files.deleteIfExists(targetManifestPath);
+        try (FileOutputStream out = new FileOutputStream(targetManifestPath.toFile())) {
```

Comment:
I might be wrong, but if the `META-INF` folder didn't exist in any of the copied jars, won't this throw an exception, because the parent of the file that is being created doesn't exist?